sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala

```diff
@@ -559,6 +559,42 @@ object StructType extends AbstractDataType {
       case _ => dt
     }
 
+  /**
+   * This works a little similarly to `merge`, but it does not actually merge two DataTypes.
+   * This method just merges nullability.
+   */
+  private[sql] def mergeNullability(left: DataType, right: DataType): DataType =
+    (left, right) match {
+      case (ArrayType(leftElementType, leftContainsNull),
+          ArrayType(rightElementType, rightContainsNull)) =>
+        ArrayType(
+          merge(leftElementType, rightElementType),
+          leftContainsNull || rightContainsNull)
+
+      case (MapType(leftKeyType, leftValueType, leftContainsNull),
+          MapType(rightKeyType, rightValueType, rightContainsNull)) =>
+        MapType(
+          merge(leftKeyType, rightKeyType),
+          merge(leftValueType, rightValueType),
+          leftContainsNull || rightContainsNull)
+
+      case (StructType(leftFields), StructType(rightFields)) =>
+        require(leftFields.size == rightFields.size, "To merge nullability," +
+          "two structs must have same number of fields.")
+
+        val newFields = leftFields.zip(rightFields).map {
+          case (leftField@StructField(_, leftType, leftNullable, _),
+              _@StructField(_, rightType, rightNullable, _)) =>
+            leftField.copy(
+              dataType = mergeNullability(leftType, rightType),
+              nullable = leftNullable || rightNullable)
+        }.toSeq
+        StructType(newFields)
+
+      case (leftType, _) =>
+        leftType
```

Comment:
Previously `merge` would do type compatibility checking:
https://github.com/apache/spark/blob/924b4130244ec73df200dc6f3b9509dd8cafba71/sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala#L663-L667
Do we need to restore this logic, or does `union` handle doing type compatibility checking elsewhere?

Reply:
`union` needs the schema of its children to be compatible. `CheckAnalysis` does the final check.

Reply:
ack, makes sense. Maybe we should mention in the Scaladoc that compatibility verification is _not_ done, since this is different from how `merge` works?